### PR TITLE
Refs #27090 -- Added real database sequence introspection

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

**Comment by claudep** [comment]
I need someone to add Oracle support, please.

**Comment by felixxm** [comment]
@claudep Currently we cannot add real database sequence introspection on Oracle, because we don't support Oracle identity fields. It will be possible after #8055 (see [_get_sequence_name](https://github.com/django/django/pull/8055/files#diff-ea23a0501b6f65fb447da98953a35dafR514)).

**Comment by claudep** [comment]
Rebased now that #8055 has been merged.

**Comment by timgraham** [comment]
@felixxm, I think we're waiting for you to do the Oracle portion of this, right?

**Comment by felixxm** [comment]
Real database sequence introspection on Oracle:
```python
def get_sequences(self, cursor, table_name):
    cursor.execute("""
        SELECT
            user_tab_identity_cols.sequence_name,
            user_tab_identity_cols.column_name
        FROM
            user_tab_identity_cols,
            user_constraints,
            user_cons_columns cols
        WHERE
            user_constraints.constraint_name = cols.constraint_name
            AND user_constraints.table_name = user_tab_identity_cols.table_name
            AND cols.column_name = user_tab_identity_cols.column_name
            AND user_constraints.constraint_type = 'P' 
            AND user_tab_identity_cols.table_name = UPPER(%s)
    """, [table_name])
    return [{'name': row[0], 'table': table_name, 'column': row[1]} for row in cursor.fetchall()]
```
We cannot rename sequence on Oracle but we can add a simple test to check if `name` is filled.

**Comment by claudep** [comment]
buildbot, test on oracle.

**Comment by claudep** [comment]
buildbot, test on oracle.

**Comment by claudep** [comment]
buildbot, test on oracle.

**Comment by claudep** [comment]
Any final review?

File: django/db/backends/mysql/introspection.py [hunk 1]
```diff
@@ -105,6 +105,13 @@ def to_int(i):
             )
         return fields
 
+    def get_sequences(self, cursor, table_name):
+        for field_info in self.get_table_description(cursor, table_name):
+            if 'auto_increment' in field_info.extra:
+                # MySQL only allows one auto increment column by table
```
Comment (timgraham) [inline django/db/backends/mysql/introspection.py]: `# MySQL allows only one auto-increment column per table.`

File: django/db/backends/oracle/introspection.py [hunk 2]
```diff
@@ -98,6 +98,28 @@ def table_name_converter(self, name):
...
+                user_constraints,
+                user_cons_columns cols
+            WHERE
+                user_constraints.constraint_name = cols.constraint_name
+                AND user_constraints.table_name = user_tab_identity_cols.table_name
+                AND cols.column_name = user_tab_identity_cols.column_name
+                AND user_constraints.constraint_type = 'P'
+                AND user_tab_identity_cols.table_name = UPPER(%s)
+        """, [table_name])
+        # Oracle allows for only one identity field per table.
```
Comment (timgraham) [inline django/db/backends/oracle/introspection.py]: chop "for"
field -> column ?

File: django/db/backends/postgresql/introspection.py [hunk 3]
```diff
@@ -82,6 +82,25 @@ def get_table_description(self, cursor, table_name):
...
+                JOIN pg_namespace sn ON sn.oid = s.relnamespace
+                JOIN pg_depend d ON d.refobjid = s.oid AND d.refclassid='pg_class'::regclass
+                JOIN pg_attrdef ad ON ad.oid = d.objid AND d.classid = 'pg_attrdef'::regclass
+                JOIN pg_attribute col ON col.attrelid = ad.adrelid AND col.attnum = ad.adnum
+                JOIN pg_class tbl ON tbl.oid = ad.adrelid
+                JOIN pg_namespace n ON n.oid = tbl.relnamespace
+            WHERE s.relkind = 'S'
+              AND d.deptype in ('a', 'n')
+              AND n.nspname = %s
+              AND tbl.relname = %s""", ['public', table_name])
```
Comment (timgraham) [inline django/db/backends/postgresql/introspection.py]: put """ on next line to match Oracle style?

Comment (timgraham) [reply django/db/backends/postgresql/introspection.py]: Is there a reason the `'public'` constant as a parameter?

File: django/db/backends/base/introspection.py [hunk 4]
```diff
@@ -113,29 +113,34 @@ def sequence_list(self):
...
 
         return sequence_list
 
+    def get_sequences(self, cursor, table_name):
+        """
+        Return a list of introspected sequences for table_name. Each sequence
+        is a dict: {'table': <table_name>, 'column': <column_name>}. An optional
+        'name' key can be added if the backend supports named sequences.
+        """
+        return []
```
Comment (timgraham) [inline django/db/backends/base/introspection.py]: Would it be better to raise `NotImplementedError()` so backends don't overlook implementing it? They can always return an empty list if it's not supported.

File: docs/releases/2.0.txt [hunk 5]
```diff
@@ -396,6 +396,10 @@ backends.
   :class:`~django.db.models.functions.Cast` function for a ``CharField`` if the
   ``max_length`` argument isn't provided.
 
+* Third-party backends should now implement the
```
Comment (timgraham) [inline docs/releases/2.0.txt]:
```
* Third-party database backends should add a
  ``DatabaseIntrospection.get_sequences()`` method based on the stub in
  ``BaseDatabaseIntrospection``.
```

File: tests/backends/postgresql/test_server_side_cursors.py [hunk 6]
```diff
@@ -80,3 +80,20 @@ def test_server_side_cursors_setting(self):
             next(persons)  # Should not open a server-side cursor
             cursors = self.inspect_cursors()
             self.assertEqual(len(cursors), 0)
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', "Test only for PostgreSQL")
+class DatabaseSequenceTests(TestCase):
+    def test_sequence_introspection(self):
```
Comment (timgraham) [inline tests/backends/postgresql/test_server_side_cursors.py]: test_get_sequences? Should other backends have similar tests? (e.g. Oracle also returns 'name')

File: tests/backends/postgresql/test_server_side_cursors.py [hunk 7]
```diff
@@ -80,3 +80,20 @@ def test_server_side_cursors_setting(self):
             next(persons)  # Should not open a server-side cursor
             cursors = self.inspect_cursors()
             self.assertEqual(len(cursors), 0)
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', "Test only for PostgreSQL")
```
Comment (timgraham) [inline tests/backends/postgresql/test_server_side_cursors.py]: Put this in a new file, `test_introspection.py`.

File: tests/backends/postgresql/test_server_side_cursors.py (hunk 6 above)
Comment (claudep) [reply tests/backends/postgresql/test_server_side_cursors.py]: This test was added in relation to the original issue which was about PostgreSQL sequences. If someone wants to add an Oracle test, why not, but I'm not working on Oracle at all.

**Comment by felixxm** [comment]
We support Oracle identity columns since Django 2.0, but there is no way to update existing columns to identity columns. Identity cannot be "attached" to existing columns, that's why we should add support also for manually created sequences. Hence if there is an `AutoField` in model's fields and `get_sequences()` return `[]`, we should return `{'table': table_name, 'column': column_name}` instead. Maybe we can pass `model._meta` instead of `table_name` to the `get_sequences()` and add something similar to the previous solution in `get_sequences()` on Oracle. 🤔
```python
row = cursor.fetchone()
if row:
    return  [{'name': row[0].lower(), 'table': model_meta.db_table, 'column': row[1].lower()}]
else:
    for f in model_meta.local_fields:
        if isinstance(f, models.AutoField):
            return [{'table': model_meta.db_table, 'column': f.column})]
```

**Comment by claudep** [comment]
If only Oracle-related issues are left in this patch, I'm fine with handing over the patch to Mariusz (or anyone else).

**Comment by claudep** [comment]
...or we can commit this now (to not miss the new feature freeze) and solve remaining Oracle concerns later in the cycle.

**Comment by felixxm** [comment]
I'll try to finish this today or tomorrow.

**Comment by felixxm** [comment]
Updated in #9069.

**Pull request closed without merging**